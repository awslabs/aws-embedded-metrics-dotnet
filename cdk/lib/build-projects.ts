import * as cdk from '@aws-cdk/core';
import { BuildSpec, Source, Project, PipelineProject, LinuxBuildImage } from '@aws-cdk/aws-codebuild';
import { PolicyStatement } from '@aws-cdk/aws-iam';

const environment = {
    // grants sudo permissions so we can use docker
    privileged: true,
    // AL2 3.0 build spec which supports .net 3.1
    // https://github.com/aws/aws-codebuild-docker-images/blob/407e02949dd24cd14a0db830c03639f34ff46ced/al2/x86_64/standard/3.0/Dockerfile#L406-L412
    buildImage: LinuxBuildImage.AMAZON_LINUX_2_3
};

export class BuildProjects {
    main: Project;
    canary: Project;
    publish: Project;

    constructor(stack: cdk.Construct) {
        this.main = new Project(stack, 'aws-embedded-metrics-dotnet', {
            projectName: 'aws-embedded-metrics-dotnet',
            buildSpec: BuildSpec.fromSourceFilename('buildspecs/buildspec.yml'),
            description: 'Default build for PRs',
            source: Source.gitHub({
                owner: 'awslabs',
                repo: 'aws-embedded-metrics-dotnet',
                webhook: true,
            }),
            environment,
        });

        this.main.addToRolePolicy(new PolicyStatement({
            actions: ['ssm:GetParameters'],
            resources: ['*']
        }));

        this.canary = new PipelineProject(stack, 'dotnet-canary', {
            projectName: 'dotnet-canary',
            buildSpec: BuildSpec.fromSourceFilename('buildspecs/buildspec.canary.yml'),
            environment,
        });

        this.canary.addToRolePolicy(new PolicyStatement({
            actions: [
                'ssm:GetParameters',
                'ecr:GetLogin',
                'ecs:UpdateService',
                'ecs:RegisterTaskDefinition',
            ],
            resources: ['*']
        }));

        this.publish = new PipelineProject(stack, 'dotnet-release', {
            projectName: 'dotnet-release',
            buildSpec: BuildSpec.fromSourceFilename('buildspecs/buildspec.release.yml'),
            environment,
        });

        this.publish.addToRolePolicy(new PolicyStatement({
            actions: [
                'ssm:GetParameters',
                'ecr:GetLogin',
                'ecs:UpdateService',
                'ecs:RegisterTaskDefinition',
            ],
            resources: ['*']
        }));
    }
}