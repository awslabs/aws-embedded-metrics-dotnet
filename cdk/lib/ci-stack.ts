import * as cdk from '@aws-cdk/core';
import { BuildSpec, Source, Project, PipelineProject, LinuxBuildImage, ComputeType } from '@aws-cdk/aws-codebuild';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import * as ecr from '@aws-cdk/aws-ecr';
import * as cw from '@aws-cdk/aws-cloudwatch';

const environment = {
  // grants sudo permissions so we can use docker
  privileged: true,
  // AL2 3.0 build spec which supports .net 3.1
  // https://github.com/aws/aws-codebuild-docker-images/blob/407e02949dd24cd14a0db830c03639f34ff46ced/al2/x86_64/standard/3.0/Dockerfile#L406-L412
  buildImage: LinuxBuildImage.AMAZON_LINUX_2_3
};

export class CIStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    this.createPipeline();
    this.createCanaryResources();
    this.createDashboard();
  }

  createPipeline() {
    const buildProject = new Project(this, 'aws-embedded-metrics-dotnet', {
      projectName: 'aws-embedded-metrics-dotnet',
      buildSpec: BuildSpec.fromSourceFilename('buildspecs/buildspec.yml'),
      description: 'Default build for PRs',
      source: Source.gitHub({
        owner: 'awslabs',
        repo: 'aws-embedded-metrics-dotnet',
        webhook: true,
      }),
      environment,
    });

    buildProject.addToRolePolicy(new PolicyStatement({
      actions: ['ssm:GetParameters'],
      resources: ['*']
    }));

    const canaryReleaseProject = new PipelineProject(this, 'dotnet-canary', {
      projectName: 'dotnet-canary',
      buildSpec: BuildSpec.fromSourceFilename('buildspecs/buildspec.canary.yml'),
      environment,
    });

    canaryReleaseProject.addToRolePolicy(new PolicyStatement({
      actions: [
        'ssm:GetParameters',
        'ecr:GetLogin',
        'ecs:UpdateService',
        'ecs:RegisterTaskDefinition',
      ],
      resources: ['*']
    }));

    const sourceOutput = new codepipeline.Artifact();
    const buildOutput = new codepipeline.Artifact();

    new codepipeline.Pipeline(this, 'dotnet-pipeline', {
      pipelineName: 'aws-embedded-metrics-dotnet-pipeline',
      stages: [
        {
          stageName: 'Source',
          actions: [
            new codepipeline_actions.GitHubSourceAction({
              actionName: 'Build',
              branch: 'main',
              owner: 'awslabs',
              repo: 'aws-embedded-metrics-dotnet',
              output: sourceOutput,
              oauthToken: cdk.SecretValue.secretsManager('github-token'),
            }),
          ],
        },
        {
          stageName: 'Build',
          actions: [
            new codepipeline_actions.CodeBuildAction({
              actionName: 'Build',
              project: buildProject,
              input: sourceOutput,
              outputs: [buildOutput],
            }),
          ],
        },
        {
          stageName: 'Deploy-Canary',
          actions: [
            new codepipeline_actions.CodeBuildAction({
              actionName: 'Deploy-Canary',
              project: canaryReleaseProject,
              input: sourceOutput,
              outputs: [new codepipeline.Artifact()],
            }),
          ],
        }
      ],
    });
  }

  createCanaryResources() {
    new ecr.Repository(this, 'emf-dotnet-canary', {
      repositoryName: 'emf-dotnet-canary'
    });
  }

  createDashboard() {
    new cw.Dashboard(this, 'Dashboard', {
      dashboardName: 'emf-dotnet-deployment',
      widgets: [
        [
          new cw.GraphWidget({
            title: 'Memory',
            width: 12,
            left: [
              new cw.MathExpression({
                expression: `SEARCH('{Canary,Runtime,Platform,Agent,Version}  MetricName="Memory.RSS" AND "Dotnet"', 'Average', 60)`,
                usingMetrics: {}
              })
            ],
            right: [
              new cw.Metric({
                namespace: 'AWS/ECS',
                metricName: 'CPUUtilization',
                dimensions: {
                  ServiceName: 'emf-dotnet-canary',
                  ClusterName: 'emf-canary'
                },
                color: '#c7c7c7'
              })
            ]
          }),
          new cw.GraphWidget({
            title: 'Init',
            width: 12,
            left: [
              new cw.MathExpression({
                expression: `SEARCH('{Canary,Runtime,Platform,Agent,Version} MetricName="Init" AND "Dotnet"', 'Sum', 60)`,
                usingMetrics: {}
              })
            ],
          }),
        ],
        [
          new cw.LogQueryWidget({
            title: 'Application Errors',
            width: 12,
            logGroupNames: ['/ecs/emf-dotnet-canary'],
            queryString: `
            filter @logStream ~= 'canary'
            | fields @timestamp, @message
            | sort @timestamp desc 
            | limit 20`,
          }),
          new cw.GraphWidget({
            title: 'Event Count',
            width: 12,
            left: [
              new cw.MathExpression({
                expression: `SEARCH('{Canary,Runtime,Platform,Agent,Version} MetricName="Invoke" AND "Dotnet"', 'SampleCount', 60)`,
                usingMetrics: {}
              })
            ],
          }),
        ],
        [
          new cw.LogQueryWidget({
            title: 'Recent EMF Data',
            width: 24,
            logGroupNames: ['/Canary/Dotnet/CloudWatchAgent/Metrics'],
            queryString: `
            fields @timestamp, @message
            | sort @timestamp desc
            | limit 20`,
          }),
        ]
      ]
    });
  }
}
