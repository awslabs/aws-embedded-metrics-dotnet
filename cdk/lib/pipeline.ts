import * as cdk from '@aws-cdk/core';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import { BuildProjects } from './build-projects';

export class Pipeline {
    stack: cdk.Construct;

    constructor(stack: cdk.Construct, buildProjects: BuildProjects) {
        const sourceOutput = new codepipeline.Artifact();
        const buildOutput = new codepipeline.Artifact();

        new codepipeline.Pipeline(stack, 'dotnet-pipeline', {
            pipelineName: 'aws-embedded-metrics-dotnet-pipeline',
            stages: [
                {
                    stageName: 'Source',
                    actions: [
                        new codepipeline_actions.GitHubSourceAction({
                            actionName: 'Build',
                            branch: 'main',
                            owner: 'awslabs',
                            repo: 'aws-embedded-metrics-dotnet',
                            output: sourceOutput,
                            oauthToken: cdk.SecretValue.secretsManager('github-token'),
                        }),
                    ],
                },
                {
                    stageName: 'Build',
                    actions: [
                        new codepipeline_actions.CodeBuildAction({
                            actionName: 'Build',
                            project: buildProjects.main,
                            input: sourceOutput,
                            outputs: [buildOutput],
                        }),
                    ],
                },
                {
                    stageName: 'Deploy-Canary',
                    actions: [
                        new codepipeline_actions.CodeBuildAction({
                            actionName: 'Deploy-Canary',
                            project: buildProjects.canary,
                            input: sourceOutput,
                            outputs: [new codepipeline.Artifact()],
                        }),
                    ],
                },
                {
                    stageName: 'Bake',
                    actions: [
                        new codepipeline_actions.ManualApprovalAction({
                            actionName: 'Bake',
                        }),
                    ],
                },
                {
                    stageName: 'Release',
                    actions: [
                        new codepipeline_actions.CodeBuildAction({
                            actionName: 'Release',
                            project: buildProjects.publish,
                            input: sourceOutput,
                            outputs: [new codepipeline.Artifact()],
                        }),
                    ],
                }
            ],
        });
    }


}