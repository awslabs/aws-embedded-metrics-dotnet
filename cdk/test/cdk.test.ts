import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Cdk from '../lib/code-build-stack';

test('Snapshot test', () => {
    const app = new cdk.App();
    const stack = new Cdk.CodeBuildStack(app, 'MyTestStack');
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
