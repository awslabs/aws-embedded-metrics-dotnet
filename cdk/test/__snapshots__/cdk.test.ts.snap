// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Resources": Object {
    "awsembeddedmetricsdotnetC8AA8DA9": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "Description": "Default build for PRs",
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "awsembeddedmetricsdotnetRoleAFA1904B",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "buildspecs/buildspec.yml",
          "Location": "https://github.com/awslabs/aws-embedded-metrics-dotnet.git",
          "ReportBuildStatus": true,
          "Type": "GITHUB",
        },
        "Triggers": Object {
          "Webhook": true,
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "awsembeddedmetricsdotnetRoleAFA1904B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "awsembeddedmetricsdotnetRoleDefaultPolicyC569B863": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "awsembeddedmetricsdotnetC8AA8DA9",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "awsembeddedmetricsdotnetC8AA8DA9",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "awsembeddedmetricsdotnetC8AA8DA9",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "awsembeddedmetricsdotnetRoleDefaultPolicyC569B863",
        "Roles": Array [
          Object {
            "Ref": "awsembeddedmetricsdotnetRoleAFA1904B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
